File: src/MoMoBot.Client.UserMedia/src/services/voiceService.js
import config from "../config";
import request from "../utils/request";
import xmlbuilder from 'xmlbuilder';
const sdk = require('microsoft-cognitiveservices-speech-sdk');

const subscriptionKey = "65789dd635a8415bbf7d7d91d0a86b76";
const serviceRegion = "eastasia";

// now create the audio-config pointing to our stream and
// the speech config specifying the language.

const speechConfig = sdk.SpeechConfig.fromSubscription(subscriptionKey, serviceRegion);

// setting the recognition language to English.
speechConfig.speechRecognitionLanguage = "zh-CN";


export const upload = (blob, duration) => {
    let formData = new FormData();
    formData.append("file", blob);
    formData.append("duration", duration)
    return request({
        url: `${config.serverUrl}/api/voice/upload`,
        method: 'POST',
        data: formData
    });
}

export const convertToText = (blob, callback) => {
    var fileReader = new FileReader();
    let pushStream = null;
    fileReader.onload = (event) => {
        pushStream = sdk.AudioInputStream.createPushStream();
        const audioConfig = sdk.AudioConfig.fromStreamInput(pushStream);
        // create the speech recognizer.
        const recognizer = new sdk.SpeechRecognizer(speechConfig, audioConfig);
        const arrayBuffer = event.target.result;
        pushStream.write(arrayBuffer);
        recognizer.recognizeOnceAsync((result) => {
            const { text = "没有听清" } = result;
            recognizer.close();
            callback(text);
            // recognizer = undefined;
        }, (err) => {
            console.error("err - " + err);
            recognizer.close();
            // recognizer = undefined;
            callback('没有听清');
        });
    };
    fileReader.onloadend = () => {
        pushStream !== null && pushStream.close();
    }
    fileReader.readAsArrayBuffer(blob);
}

export const convertToVoice = (text,True) => {
    // let options = {
    //     method: 'POST',
    //     url: `https://${serviceRegion}.api.cognitive.microsoft.com/sts/v1.0/issueToken`,
    //     headers: {
    //         'Ocp-Apim-Subscription-Key': subscriptionKey
    //     }
    // };
    // return request(options)
    //     .then(accessToken => {
    //         let xml_body = `<speak version='1.0' xmlns='http://www.w3.org/2001/10/synthesis' xml:lang='zh-CN'>
    //         <voice name='Microsoft Server Speech Text to Speech Voice (zh-CN, Kangkang, Apollo)'>"${text}"</voice></speak>`
    //         // Convert the XML into a string to send in the TTS request.
    //         let body = xml_body.toString();
    //         request({
    //             method: 'POST',
    //             baseUrl: '',
    //             url: `https://${serviceRegion}.tts.speech.microsoft.com/cognitiveservices/v1`,
    //             headers: {
    //                 'Authorization': 'Bearer ' + accessToken,
    //                 'cache-control': 'no-cache',
    //                 // 'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1',
    //                 'X-Microsoft-OutputFormat': 'riff-24khz-16bit-mono-pcm',
    //                 'Content-Type': 'application/ssml+xml'
    //             },
    //             data : body
    //         }).then(response=>{
    //             True(response);
    //         })
    //     })
}

export const getVoiceFile = (id) => {
    return request({
        url: `${config.serverUrl}/api/voice/${id}`,
        method: 'GET'
    });
}

File: src/MoMoBot.Service/Implements/VoiceService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using MoMoBot.Infrastructure.Database;
using MoMoBot.Infrastructure.Models;
using MoMoBot.Service.Abstractions;

namespace MoMoBot.Service.Implements
{
    public class VoiceService
        : IVoiceService
    {
        private readonly MoMoDbContext _dbContext;
        private readonly string _voicesPath;
        private readonly Authentication _authentication;
        private readonly string _abstractPath;
        private readonly HttpClient _http;
        public VoiceService(IHostingEnvironment env, IConfiguration config, MoMoDbContext dbContext, HttpClient http)
        {
            _http = http;
            _dbContext = dbContext;
            var endpoint = config.GetValue("CognitiveServicesKeys:Endpoint", "");
            var key = config.GetValue("CognitiveServicesKeys:Voice", "");
            _authentication = new Authentication($"{endpoint}/sts/v1.0/issueToken", key);
            _voicesPath = Path.Combine("voices", DateTime.Now.ToString("yyyyMMdd"));
            _abstractPath = Path.Combine(env.ContentRootPath, "upload", _voicesPath);
            if (Directory.Exists(_abstractPath) == false)
            {
                Directory.CreateDirectory(_abstractPath);
            }
        }

        public async Task<Guid> CreateAsync(IFormFile file, double duration, string type = "wav")
        {
            if (file != null)
            {
                var id = Guid.NewGuid();

                var filename = $"{id}.wav";
                var path = Path.Combine(_voicesPath, filename);
                var abstrctPath = Path.Combine(_abstractPath, filename);

                using (var writer = new FileStream(abstrctPath, FileMode.OpenOrCreate, FileAccess.ReadWrite))
                {
                    await file.CopyToAsync(writer);
                }
                await _dbContext.Voices.AddAsync(new Voice
                {
                    CreationTime = DateTime.Now,
                    AudioType = type,
                    Id = id,
                    Duration = duration,
                    SavePath = path
                });
                await _dbContext.SaveChangesAsync();
                return id;
            }
            return Guid.Empty;
        }



        public async Task<Voice> GetAsync(Guid id)
        {
            var voice = await _dbContext.Voices.FirstOrDefaultAsync(v => v.Id == id);
            return voice;
        }

        public async Task<string> TTS(string text)
        {
            if (!string.IsNullOrEmpty(text))
            {
                return await GetOrCreateAsync(text);
            }

            return string.Empty;
        }

        private async Task<string> GetOrCreateAsync(string text)
        {
            var path = string.Empty;
            var voice = await _dbContext.Voices.FirstOrDefaultAsync(v => v.Text.Equals(text, StringComparison.CurrentCultureIgnoreCase));
            if (voice != null)
            {
                path = voice.SavePath;
            }
            else
            {
                var accessToken = await _authentication.FetchTokenAsync();
                var host = "https://southeastasia.tts.speech.microsoft.com/cognitiveservices/v1";
                var body = @"<speak version='1.0' xmlns='http://www.w3.org/2001/10/synthesis' xml:lang='zh-CN'>
              <voice name='Microsoft Server Speech Text to Speech Voice (zh-CN, XiaoxiaoNeural)'>" +
                  text + "</voice></speak>";

                using (var request = new HttpRequestMessage())
                {
                    // Set the HTTP method
                    request.Method = HttpMethod.Post;
                    // Construct the URI
                    request.RequestUri = new Uri(host);
                    // Set the content type header
                    request.Content = new StringContent(body, Encoding.UTF8, "application/ssml+xml");
                    // Set additional header, such as Authorization and User-Agent
                    request.Headers.Add("Authorization", "Bearer " + accessToken);
                    request.Headers.Add("Connection", "Keep-Alive");
                    // Update your resource name
                    request.Headers.Add("User-Agent", "Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1");
                    request.Headers.Add("X-Microsoft-OutputFormat", "riff-24khz-16bit-mono-pcm");
                    // Create a request
                    using (var response = await _http.SendAsync(request).ConfigureAwait(false))
                    {
                        response.EnsureSuccessStatusCode();
                        // Asynchronously read the response
                        using (var dataStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
                        {
                            var id = Guid.NewGuid();
                            var filename = $"{id}.wav";
                            path = Path.Combine(_voicesPath, filename);
                            var abstrctPath = Path.Combine(_abstractPath, filename);
                            using (var writer = new FileStream(abstrctPath, FileMode.OpenOrCreate, FileAccess.ReadWrite))
                            {
                                await dataStream.CopyToAsync(writer);
                            }
                            await _dbContext.Voices.AddAsync(new Voice
                            {
                                Text = text,
                                CreationTime = DateTime.Now,
                                AudioType = "wav",
                                Id = id,
                                Duration = 0d,
                                SavePath = path
                            });
                            await _dbContext.SaveChangesAsync();
                        }
                    }
                }

            }

            return Path.Combine("files", path);
        }
    }

    public class Authentication
    {
        private string subscriptionKey;
        private string tokenFetchUri;

        public Authentication(string tokenFetchUri, string subscriptionKey)
        {
            if (string.IsNullOrWhiteSpace(tokenFetchUri))
            {
                throw new ArgumentNullException(nameof(tokenFetchUri));
            }
            if (string.IsNullOrWhiteSpace(subscriptionKey))
            {
                throw new ArgumentNullException(nameof(subscriptionKey));
            }
            this.tokenFetchUri = tokenFetchUri;
            this.subscriptionKey = subscriptionKey;
        }

        public async Task<string> FetchTokenAsync()
        {
            using (var client = new HttpClient())
            {
                client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", this.subscriptionKey);
                UriBuilder uriBuilder = new UriBuilder(this.tokenFetchUri);

                var result = await client.PostAsync(uriBuilder.Uri.AbsoluteUri, null).ConfigureAwait(false);
                return await result.Content.ReadAsStringAsync().ConfigureAwait(false);
            }
        }
    }
}
